Mathlib/NumberTheory/NumberField/Basic.lean

```diff
@@ -65,34 +65,66 @@ instance [NumberField L] [Algebra K L] : FiniteDimensional K L :=
 
 /-- The ring of integers (or number ring) corresponding to a number field
 is the integral closure of ℤ in the number field. -/
-def ringOfIntegers :=
+def RingOfIntegers : Type _ :=
   integralClosure ℤ K
-#align number_field.ring_of_integers NumberField.ringOfIntegers
+#align number_field.ring_of_integers NumberField.RingOfIntegers
 
-@[inherit_doc] scoped notation "𝓞" => NumberField.ringOfIntegers
+@[inherit_doc] scoped notation "𝓞" => NumberField.RingOfIntegers
 
-theorem mem_ringOfIntegers (x : K) : x ∈ 𝓞 K ↔ IsIntegral ℤ x :=
-  Iff.rfl
-#align number_field.mem_ring_of_integers NumberField.mem_ringOfIntegers
+namespace RingOfIntegers
 
-theorem isIntegral_of_mem_ringOfIntegers {K : Type*} [Field K] {x : K} (hx : x ∈ 𝓞 K) :
-    IsIntegral ℤ (⟨x, hx⟩ : 𝓞 K) := by
-  obtain ⟨P, hPm, hP⟩ := hx
-  refine' ⟨P, hPm, _⟩
-  rw [← Polynomial.aeval_def, ← Subalgebra.coe_eq_zero, Polynomial.aeval_subalgebra_coe,
-    Polynomial.aeval_def, Subtype.coe_mk, hP]
-#align number_field.is_integral_of_mem_ring_of_integers NumberField.isIntegral_of_mem_ringOfIntegers
+instance : CommRing (𝓞 K) :=
+  inferInstanceAs (CommRing (integralClosure _ _))
+instance : IsDomain (𝓞 K) :=
+  inferInstanceAs (IsDomain (integralClosure _ _))
+instance : CharZero (𝓞 K) :=
+  inferInstanceAs (CharZero (integralClosure _ _))
+instance : Algebra (𝓞 K) K :=
+  inferInstanceAs (Algebra (integralClosure _ _) _)
+instance : NoZeroSMulDivisors (𝓞 K) K :=
+  inferInstanceAs (NoZeroSMulDivisors (integralClosure _ _) _)
+
+variable {K}
+
+@[ext] theorem ext {x y : 𝓞 K} (h : algebraMap _ K x = algebraMap _ K y) : x = y :=
+  Subtype.ext h
+theorem ext_iff {x y : 𝓞 K} : x = y ↔ algebraMap _ K x = algebraMap _ K y :=
+  Subtype.ext_iff
+
+@[simp] lemma map_mk (x : K) (hx) : algebraMap (𝓞 K) K ⟨x, hx⟩ = x := rfl
+
+lemma mk_eq_mk (x y : K) (hx hy) : (⟨x, hx⟩ : 𝓞 K) = ⟨y, hy⟩ ↔ x = y := by simp
+
+@[simp] lemma mk_one : (⟨1, one_mem _⟩ : 𝓞 K) = 1 :=
+  rfl
+
+@[simp] lemma mk_zero : (⟨0, zero_mem _⟩ : 𝓞 K) = 0 :=
+  rfl
+-- TODO: these lemmas don't seem to fire?
```

Comment:
We should figure out why `simp` dislikes these lemmas.

Reply:
I tried for a bit and can't figure it out. If I write `mk_add_mk _` then `simp` can use it (e.g. in `restrict_addMonoidHom`), but `mk_add_mk` is too hard.

Reply:
Weird

Reply:
If it works with `_` and not without it, then there are some instance implicit arguments that cannot be found (or unified correctly) with typeclass synthesis. The underscore bypasses typeclass search for all arguments (if I remember correctly).

Reply:
It's not because of the `⟨x, hx⟩` syntax?

Reply:
> If it works with `_` and not without it, then there are some instance implicit arguments that cannot be found (or unified correctly) with typeclass synthesis. The underscore bypasses typeclass search for all arguments (if I remember correctly).

I forgot to give more details but they seem not to even be considered according to the `set_option trace` output. So it doesn't even appear to arrive at the instance synth step. But afaict the terms match perfectly syntactically: the discrimination tree should just find the simp lemmas.